drivers/i2s/i2s_ll_stm32.c

```diff
@@ -349,9 +359,30 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
 			LOG_ERR("DRAIN trigger: invalid state");
 			return -EIO;
 		}
-		stream->stream_disable(stream, dev);
-		stream->queue_drop(stream);
-		stream->state = I2S_STATE_READY;
+
+		if (dir == I2S_DIR_TX) {
+			if (stream->state == I2S_STATE_RUNNING) {
+				stream->state = I2S_STATE_STOPPING;
+				/*
+				 * Indicate that the transition to I2S_STATE_STOPPING
+				 * is triggered by DRAIN command
+				 */
+				stream->sample_tx_stop_drain = true;
+			} else if ((queue_is_empty(&stream->mem_block_queue) == false) ||
+						(ll_func_i2s_dma_busy(cfg->i2s))) {
+				stream->stream_disable(stream, dev);
+				stream->queue_drop(stream);
+				stream->state = I2S_STATE_READY;
+			} else
+				return -EIO;
+		}
+
+		if (dir == I2S_DIR_RX) {
```

Comment:
Please use `else if`.

Nit: replacing this code path with a `goto do_trigger_stop`, where `trigger_stop` is a label added in the `case I2S_TRIGGER_STOP` block above, could be cleaner, as it reduces code duplication and is more in line with the Zephyr API documentation:
```
If the [I2S_TRIGGER_DRAIN] trigger is applied to the RX queue it has the same effect as I2S_TRIGGER_STOP.
``` 

Nit 2: I2S_TRIGGER_STOP is currently broken - maybe a comment should be added to indicate this?